### DISCUSSION TITLE
How to associate data with an API session?

### DISCUSSION THREAD

**Discussion by notramo** [luckyframework/lucky#1809] [Q&A] [answered]
I want to provide different data on given routes, based on parameters provided at login time. How to persist this data to the session?

It seems like it's possible to accomplish this functionality by putting the data into the JWT token, but is there an idiomatic way to access it inside controller/templates?

**Comment by jwoertink** [comment]
There's a Lucky shard that lets you store session in redis, otherwise you're at the mercy of cookie limits with normal sessions. But provided your data doesn't get too crazy, just use session.set(:data, jwt)

**Reply by notramo** [reply]
It's an API server, that doesn't have cookies. I examined the current JWT-based implementation, and it seems that storing it in the JWT token would be a viable solution, since the value is known at login time, and doesn't change afterwards. Is there an idiomatic way of accessing it using an interface, or do I have to patch it?

**Reply by jwoertink** [reply]
Ah, if you're doing an API, then generally what you would do is send that JWT to the client, and have the client send it back through the headers. Otherwise, I think just storing it in the DB is fine, and pull it out when you need it.

**Reply by notramo** [reply]
How to add a field to the JWT token? E.g. I want to add a `scope` field next to the `user_id`.

**Reply by jwoertink** [reply]
Assuming you're using the `UserToken` for all this, you can just do that here...

https://github.com/luckyframework/lucky_cli/blob/7c3c69c775ba9aad7ac0206c052db66a906941c1/src/api_authentication_app_skeleton/src/models/user_token.cr#L7

```crystal
payload = {"user_id" => user.id, "scope" => "something"}
```

**Reply by notramo** [reply]
Does it mean that I have to patch the current implementation?
Then how can I get the value? Do I have to patch that too? Or is there a way to get a field from the token?

**Reply by jwoertink** [reply, marked as answer]
I'm having to make a lot of assumptions here since I don't have any of your code... but assuming you're using the `UserToken` class, then it's just in your `src/models/user_token.cr` file. Just open that file and make whatever changes you need to.
